### PR TITLE
Define v1alpha1 Run type and generated scaffolding

### PR DESCRIPTION
**Pull request by ImJasonH**
This defines the basic Run type, which will be used to specify
executions of Custom Tasks and report status of those executions.

This change doesn't wire the new type in to any controllers or
validating webhooks, so that this time nothing will happen when a Run is
created. It also doesn't wire in to PipelineRun execution to enable
Pipelines to specify Custom Tasks.

This is groundwork to enable [TEP 0002](https://github.com/tektoncd/community/pull/128) and enable further experimentation. This type may change in breaking ways as a result of that experimentation and further iteration on the design.

Next steps: moar samples!!, webhook-validate Runs, enforce initial update timeout in controller, create Runs from PipelineRuns, omgsomanytests

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [y] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [n/a] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [y] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [x] Release notes block has been filled in or deleted (only if no user facing changes)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

# Release Notes

```release-note
Define a new Run type to describe executions of Custom Tasks
```

### CODE REVIEW

File: config/300-run.yaml [hunk 1]
```diff
@@ -0,0 +1,71 @@
+# Copyright 2019 The Tekton Authors
```
Comment (afrittoli) [inline config/300-run.yaml:1]: s/2019/2020

File: pkg/apis/pipeline/v1alpha1/run_defaults.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
+/*
+Copyright 2019 The Tekton Authors
```
Comment (afrittoli) [inline pkg/apis/pipeline/v1alpha1/run_defaults.go:2]: s/2019/202

File: pkg/controller/filter.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
+/*
+Copyright 2019 The Tekton Authors
```
Comment (afrittoli) [inline pkg/controller/filter.go:2]: s/2019/2020

**Comment by ImJasonH** [comment]
/retest

🙄

File: pkg/apis/pipeline/v1alpha1/run_types.go [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+	}
+)
+
+// RunSpec defines the desired state of Run
+type RunSpec struct {
+	// +optional
+	Ref *TaskRef `json:"ref,omitempty"`
+
+	// +optional
+	Params []v1beta1.Param `json:"params,omitempty"`
```
[the remark this answers is not in the document]
Comment (ImJasonH) [reply pkg/apis/pipeline/v1alpha1/run_types.go:45]: It's possible. I'm not sure how `util/set` marshals/unmarshals as JSON/YAML. If we make this change I'd like to make it across all the places we use `[]Param` at once, for uniformity.

[the remark this answers is not in the document]
Comment (ImJasonH) [reply pkg/apis/pipeline/v1alpha1/run_types.go:45]: Looking into it a bit more, [`util/sets`](https://godoc.org/k8s.io/apimachinery/pkg/util/sets) doesn't have any builtin type for slices-of-structs, only basic types, so we'd have to define it ourselves. If we think it's worth it we can do that, but it won't be as easy as adopting an already-implemented container type.

File: pkg/apis/pipeline/v1alpha1/run_types.go [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+
+	// TODO(https://github.com/tektoncd/community/pull/128)
+	// - cancellation
+	// - timeout
+	// - inline task spec
+	// - workspaces ?
+}
+
+// TODO(jasonhall): Move this to a Params type so other code can use it?
```
Comment (vdemeester) [inline pkg/apis/pipeline/v1alpha1/run_types.go:54]: :+1:

**Comment by bobcatfish** [comment]
@afrittoli mentioned this in slack but before merging I would have liked to include docs explaining these new fields - with our goal to be always in a releasable state, we'd want to make sure that if for some reason we cut a release tomorrow, there'd be an adequate explanation for users of why these types were added and do nothing ( @pritidesai did a great job of this with https://github.com/tektoncd/pipeline/pull/2650)

Also kinda related to the 2 reviewer conversation in slack, our policy for additive API changes is that they be approved by at least 2 OWNERS (https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes). Maybe jason + andrea counts, but I'd really like to get more eyes on API changes in general (selfishly: my own :D)

**Comment by bobcatfish** [comment]
> Not blocking this PR anyways.

i think the docs _should_ block the PR (in the spirit of being always releasable)

File: pkg/apis/pipeline/controller.go [hunk 6]
```diff
@@ -21,6 +21,9 @@ const (
 	// nolint: golint
 	PipelineRunControllerName = "PipelineRun"
 
-	// TaskRunControllerName holds the name of the PipelineRun controller
+	// TaskRunControllerName holds the name of the TaskRun controller
 	TaskRunControllerName = "TaskRun"
+
+	// TaskRunControllerName holds the name of the PipelineRun controller
```
Comment (bobcatfish) [inline pkg/apis/pipeline/controller.go:27]: i think the comment is a copy pasta error

File: pkg/apis/pipeline/controller.go [hunk 7]
```diff
@@ -21,6 +21,9 @@ const (
...
+	RunControllerName = "Run"
```
Comment (bobcatfish) [inline pkg/apis/pipeline/controller.go:28]: what does run controller name mean here?

File: pkg/apis/pipeline/v1alpha1/run_types.go [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+
+var (
+	runGroupVersionKind = schema.GroupVersionKind{
+		Group:   SchemeGroupVersion.Group,
+		Version: SchemeGroupVersion.Version,
+		Kind:    pipeline.RunControllerName,
+	}
+)
+
+// RunSpec defines the desired state of Run
```
Comment (bobcatfish) [inline pkg/apis/pipeline/v1alpha1/run_types.go:39]: can this docstring and the others in this file elaborate on what a Run is?

File: pkg/apis/pipeline/v1alpha1/run_types.go [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+func (rs RunSpec) GetParam(name string) *v1beta1.Param {
+	for _, p := range rs.Params {
+		if p.Name == name {
+			return &p
+		}
+	}
+	return nil
+}
+
+type RunStatus struct {
```
Comment (bobcatfish) [inline pkg/apis/pipeline/v1alpha1/run_types.go:64]: can has docstrings for all exported types plz

File: pkg/apis/pipeline/v1alpha1/run_types.go [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+
+// GetStatus retrieves the status of the Parallel. Implements the KRShaped
+// interface.
+func (r *Run) GetStatus() *duckv1.Status { return &r.Status.Status }
+
+// RunStatusFields holds the fields of Run's status.  This is defined
+// separately and inlined so that other types can readily consume these fields
+// via duck typing.
+type RunStatusFields struct {
+	// StartTime is the time the build is actually started.
```
Comment (bobcatfish) [inline pkg/apis/pipeline/v1alpha1/run_types.go:116]: some copy pasta in these i think (our old friend build)

File: pkg/apis/pipeline/v1alpha1/run_validation.go [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+	"context"
+
+	"github.com/tektoncd/pipeline/pkg/apis/validate"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"knative.dev/pkg/apis"
+)
+
+var _ apis.Validatable = (*Run)(nil)
+
+// Validate taskrun
```
Comment (bobcatfish) [inline pkg/apis/pipeline/v1alpha1/run_validation.go:29]: copy pasta?

**Pull request merged**